# Capstone Project

The project is to explore, segment, and cluster the neighborhoods in the city of Toronto

## Importing Data

In [1]:
# importing libraries

import requests
import lxml.html as lh
import pandas as pd


In [2]:
# getting the data from the website

url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
#Create a handle, page, to handle the contents of the website
page = requests.get(url)
#Store the contents of the website under doc
doc = lh.fromstring(page.content)
#Parse data that are stored between <tr>..</tr> of HTML
tr_elements = doc.xpath('//tr')

In [3]:
tr_elements[0:5]

[<Element tr at 0x2e1d39aec78>,
 <Element tr at 0x2e1d39aed68>,
 <Element tr at 0x2e1d39aeea8>,
 <Element tr at 0x2e1d39aeef8>,
 <Element tr at 0x2e1d39aee58>]

In [4]:
tr_elements = doc.xpath('//tr')
#Create empty list
col=[]
i=0
#For each row, store each first element (header) and an empty list
for t in tr_elements[0]:
    i+=1
    name=t.text_content()
    print('%d:%s'%(i,name))
    col.append((name,[]))

1:Postal code

2:Borough

3:Neighborhood



In [5]:
#Since out first row is the header, data is stored on the second row onwards
for j in range(1,len(tr_elements)):
    #T is our j'th row
    T=tr_elements[j]
    
    #If row is not of size 10, the //tr data is not from our table 
    if len(T)!=3:
        break
    
    #i is the index of our column
    i=0
    
    #Iterate through each element of the row
    for t in T.iterchildren():
        data=t.text_content() 
        #Check if row is empty
        if i>0:
        #Convert any numerical value to integers
            try:
                data=int(data)
            except:
                pass
        #Append the data to the empty list of the i'th column
        col[i][1].append(data)
        #Increment i for the next column
        i+=1

In [6]:
[len(C) for (title,C) in col]

[181, 181, 181]

In [7]:
Dict={title:column for (title,column) in col}
df=pd.DataFrame(Dict)

In [8]:
df.head()

Postal code\n           Borough\n                Neighborhood\n
0         M1A\n      Not assigned\n                            \n
1         M2A\n      Not assigned\n                            \n
2         M3A\n        North York\n                   Parkwoods\n
3         M4A\n        North York\n            Victoria Village\n
4         M5A\n  Downtown Toronto\n  Regent Park / Harbourfront\n

## Data Cleaning

In [9]:
df1 = df.replace('\n','', regex=True)

In [10]:
df1.head()

Postal code\n         Borough\n              Neighborhood\n
0           M1A      Not assigned                            
1           M2A      Not assigned                            
2           M3A        North York                   Parkwoods
3           M4A        North York            Victoria Village
4           M5A  Downtown Toronto  Regent Park / Harbourfront

In [11]:
# Renaming the columns

df1.rename(columns = {'Postal code\n' : 'Postal code', 'Borough\n' : 'Borough', 'Neighborhood\n' : 'Neighborhood' }, inplace = True)

In [12]:
# Checking first 5 rows of a dataframe

df1.head()

Postal code           Borough                Neighborhood
0         M1A      Not assigned                            
1         M2A      Not assigned                            
2         M3A        North York                   Parkwoods
3         M4A        North York            Victoria Village
4         M5A  Downtown Toronto  Regent Park / Harbourfront

In [13]:
# Checking the shape of a dataframe
df1.shape

(181, 3)

In [14]:
#Checking how many rows are Not Assigned in Borough Column

df1[df1.Borough == 'Not assigned'].count()

Postal code     77
Borough         77
Neighborhood    77
dtype: int64

#### Dropping the rows with the column 'Borough' having 'Not Assigned'

In [15]:
df1.drop(df1[df1['Borough'] == 'Not assigned'].index, inplace = True)

In [16]:
# Checking the shape of a dataframe

df1.head()

Postal code           Borough                                  Neighborhood
2         M3A        North York                                     Parkwoods
3         M4A        North York                              Victoria Village
4         M5A  Downtown Toronto                    Regent Park / Harbourfront
5         M6A        North York             Lawrence Manor / Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government

#### Checking the null values for all the 3 columns

In [17]:
df1['Postal code'].isnull().sum()

0

In [18]:
df1['Borough'].isnull().sum()

0

In [19]:
df1['Neighborhood'].isna().sum()

0

In [20]:
df1.isna().sum()

Postal code     0
Borough         0
Neighborhood    0
dtype: int64

In [21]:
df1

Postal code                Borough  \
2           M3A             North York   
3           M4A             North York   
4           M5A       Downtown Toronto   
5           M6A             North York   
6           M7A       Downtown Toronto   
..          ...                    ...   
165         M4Y       Downtown Toronto   
168         M7Y           East Toronto   
169         M8Y              Etobicoke   
178         M8Z              Etobicoke   
180              Canadian postal codes   

                                          Neighborhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                           Regent Park / Harbourfront  
5                    Lawrence Manor / Lawrence Heights  
6         Queen's Park / Ontario Provincial Government  
..                                                 ...  
165                               Church and Wellesley  
168              Business reply mail Processing CentrE  
169  Old Mill South / King's Mill Park / Sunnylea /...  
178  Mimico NW / The Queensway West / South of Bloo...  
180                                                     

[104 rows x 3 columns]

#### Checking whether the column Neighborhood has any null values or assinged to the value 'Not Assigned'

In [22]:
df1[df1['Neighborhood'] == 'Not Assigned']

Empty DataFrame
Columns: [Postal code, Borough, Neighborhood]
Index: []

In [23]:
df1['Neighborhood'].unique()

array(['Parkwoods', 'Victoria Village', 'Regent Park / Harbourfront',
       'Lawrence Manor / Lawrence Heights',
       "Queen's Park / Ontario Provincial Government", 'Islington Avenue',
       'Malvern / Rouge', 'Don Mills', 'Parkview Hill / Woodbine Gardens',
       'Garden District, Ryerson', 'Glencairn',
       'West Deane Park / Princess Gardens / Martin Grove / Islington / Cloverdale',
       'Rouge Hill / Port Union / Highland Creek', 'Woodbine Heights',
       'St. James Town', 'Humewood-Cedarvale',
       'Eringate / Bloordale Gardens / Old Burnhamthorpe / Markland Wood',
       'Guildwood / Morningside / West Hill', 'The Beaches',
       'Berczy Park', 'Caledonia-Fairbanks', 'Woburn', 'Leaside',
       'Central Bay Street', 'Christie', 'Cedarbrae', 'Hillcrest Village',
       'Bathurst Manor / Wilson Heights / Downsview North',
       'Thorncliffe Park', 'Richmond / Adelaide / King',
       'Dufferin / Dovercourt Village', 'Scarborough Village',
       'Fairview / Henry Far

In [24]:
df1[df1['Neighborhood'] == '']

Postal code                Borough Neighborhood
180              Canadian postal codes

In [25]:
# Dropping the unneccessary row

df1.drop(df1[df1['Borough'] == 'Canadian postal codes'].index, inplace = True)

In [26]:
df1.shape

(103, 3)

In [27]:
df1[df1['Neighborhood'] == '']

Empty DataFrame
Columns: [Postal code, Borough, Neighborhood]
Index: []

The column 'Neighborhood' doesn't have any 'Not Assigned'or Null values

In [28]:
df1

Postal code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                          Neighborhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                           Regent Park / Harbourfront  
5                    Lawrence Manor / Lawrence Heights  
6         Queen's Park / Ontario Provincial Government  
..                                                 ...  
160   The Kingsway / Montgomery Road  / Old Mill North  
165                               Church and Wellesley  
168              Business reply mail Processing CentrE  
169  Old Mill South / King's Mill Park / Sunnylea /...  
178  Mimico NW / The Queensway West / South of Bloo...  

[103 rows x 3 columns]

#### Checking the Duplicates

In [29]:
df1['Postal code'].duplicated().sum()

0

In [30]:
df1.duplicated().sum()

0

#### Replacing the character '/' with ','

In [31]:
toronto_data = df1.replace(' / ',', ', regex=True)

In [32]:
toronto_data.head()

Postal code           Borough                                 Neighborhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [33]:
toronto_data.reset_index(drop = True, inplace = True)

In [34]:
toronto_data.head()

Postal code           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [35]:
toronto_data.shape

(103, 3)

After cleaning the data, dataframe has 103 rows and 3 columns

Now that you have built a dataframe of the postal code of each neighborhood along with the borough name and neighborhood name, in order to utilize the Foursquare location data, we need to get the latitude and the longitude coordinates of each neighborhood.

In an older version of this course, we were leveraging the Google Maps Geocoding API to get the latitude and the longitude coordinates of each neighborhood. However, recently Google started charging for their API: http://geoawesomeness.com/developers-up-in-arms-over-google-maps-api-insane-price-hike/, so we will use the Geocoder Python package instead: https://geocoder.readthedocs.io/index.html.

The problem with this Package is you have to be persistent sometimes in order to get the geographical coordinates of a given postal code. So you can make a call to get the latitude and longitude coordinates of a given postal code and the result would be None, and then make the call again and you would get the coordinates. So, in order to make sure that you get the coordinates for all of our neighborhoods, you can run a while loop for each postal code. Taking postal code M5G as an example, your code would look something like this:

Given that this package can be very unreliable, in case you are not able to get the geographical coordinates of the neighborhoods using the Geocoder package, here is a link to a csv file that has the geographical coordinates of each postal code: http://cocl.us/Geospatial_data

In [36]:
#Importing the geospatial data

geospatial_data = pd.read_csv('Geospatial_Coordinates.csv')

In [37]:
geospatial_data.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [38]:
# Renaming the column

geospatial_data.rename(columns = {'Postal Code' : 'Postal code'},inplace = True)

In [39]:
# Merging the two dataframes on Postal Code

toronto_spatial = pd.merge(toronto_data, geospatial_data[['Postal code', 'Latitude', 'Longitude']], on='Postal code')
toronto_spatial.head()

Postal code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

In [40]:
# Checking the shape of the new dataframe

toronto_spatial.shape

(103, 5)

In [41]:
# Checking the null values of a new dataframe

toronto_spatial.isna().sum()

Postal code     0
Borough         0
Neighborhood    0
Latitude        0
Longitude       0
dtype: int64

In [42]:
# Checking the duplicates of a new dataframe

toronto_spatial.duplicated().sum()

0

Explore and cluster the neighborhoods in Toronto. You can decide to work with only boroughs that contain the word Toronto and then replicate the same analysis we did to the New York City data

In [43]:
toronto_spatial['Borough'].unique()

array(['North York', 'Downtown Toronto', 'Etobicoke', 'Scarborough',
       'East York', 'York', 'East Toronto', 'West Toronto',
       'Central Toronto', 'Mississauga'], dtype=object)

In [44]:
toronto = toronto_spatial[toronto_spatial['Borough'].str.contains('Toronto')].reset_index(drop = True)
toronto.head()

Postal code           Borough                                 Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.676357 -79.293031

In [45]:
toronto['Borough'].unique()

array(['Downtown Toronto', 'East Toronto', 'West Toronto',
       'Central Toronto'], dtype=object)

In [46]:
pip install geopy

Note: you may need to restart the kernel to use updated packages.


In [47]:
from geopy.geocoders import Nominatim

In [48]:
address = 'Toronto, Canada'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


In [49]:
import folium

In [50]:
# create map of Manhattan using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(toronto['Latitude'], toronto['Longitude'], toronto['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Define Foursquare Credentials and Version

In [51]:
CLIENT_ID = 'DNJ41VQLTGWQ40IMXM0OZVEB2DCHNHRWSWRSTHXXRIC2L4YG' # your Foursquare ID
CLIENT_SECRET = 'N3NBIBW4BM42HWLZI54V24EEPTFYF0T51PCVXDND5MQIT2B3' # your Foursquare Secret
VERSION = '20200502' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: DNJ41VQLTGWQ40IMXM0OZVEB2DCHNHRWSWRSTHXXRIC2L4YG
CLIENT_SECRET:N3NBIBW4BM42HWLZI54V24EEPTFYF0T51PCVXDND5MQIT2B3


#### Let's explore the first neighborhood in our dataframe

In [52]:
toronto.loc[0, 'Neighborhood']

'Regent Park, Harbourfront'

In [53]:
neighborhood_latitude = toronto.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = toronto.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = toronto.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Regent Park, Harbourfront are 43.6542599, -79.3606359.


#### Now, let's get the top 100 venues that are in Regent Park and Harbour Front with a radius of 500 meters

In [54]:
# type your answer here

LIMIT = 100 # limit of number of venues returned by Foursquare API

radius = 500 # define radius

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?&client_id=DNJ41VQLTGWQ40IMXM0OZVEB2DCHNHRWSWRSTHXXRIC2L4YG&client_secret=N3NBIBW4BM42HWLZI54V24EEPTFYF0T51PCVXDND5MQIT2B3&v=20200502&ll=43.6542599,-79.3606359&radius=500&limit=100'

In [55]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5eadb1281e152c001b48819b'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Corktown',
  'headerFullLocation': 'Corktown, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 48,
  'suggestedBounds': {'ne': {'lat': 43.6587599045, 'lng': -79.3544279001486},
   'sw': {'lat': 43.6497598955, 'lng': -79.36684389985142}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '54ea41ad498e9a11e9e13308',
       'name': 'Roselle Desserts',
       'location': {'address': '362 King St E',
        'crossStreet': 'Trinity St',
        'lat': 43.653446723052674,
        'lng': -79.3620167174383,
        'labeledLatLngs': [{'label': 'display',
 

In [56]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [57]:
from pandas.io.json import json_normalize

venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

C:\Users\SOWMYA\anaconda3\lib\site-packages\ipykernel_launcher.py:5: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  """


name           categories        lat        lng
0        Roselle Desserts               Bakery  43.653447 -79.362017
1           Tandem Coffee          Coffee Shop  43.653559 -79.361809
2      Morning Glory Cafe       Breakfast Spot  43.653947 -79.361149
3  Cooper Koo Family YMCA  Distribution Center  43.653249 -79.358008
4     Body Blitz Spa East                  Spa  43.654735 -79.359874

In [58]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

48 venues were returned by Foursquare.


### Explore each neigborhood

In [59]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

48 venues were returned by Foursquare.


In [60]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])
    
    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)
        
        
        
        
        
        
        

In [61]:
# type your answer here

toronto_venues = getNearbyVenues(names=toronto['Neighborhood'],
                                   latitudes=toronto['Latitude'],
                                   longitudes=toronto['Longitude']
                                  )

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West
High Park, The Junction South
North Toronto West
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Rosedale
Stn A PO Boxes
St. James Town,

Let's check how many venues were returned for each neighborhood

In [62]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            57   
Brockton, Parkdale Village, Exhibition Place                           23   
Business reply mail Processing CentrE                                  18   
CN Tower, King and Spadina, Railway Lands, Harb...                     17   
Central Bay Street                                                     63   
Christie                                                               17   
Church and Wellesley                                                   78   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             36   
Davisville North                                                        8   
Dufferin, Dovercourt Village                                           19   
First Canadian Place, Underground city                                100   
Forest Hill North & West                                                5   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          23   
India Bazaar, The Beaches West                                         20   
Kensington Market, Chinatown, Grange Park                              55   
Lawrence Park                                                           4   
Little Portugal, Trinity                                               41   
Moore Park, Summerhill East                                             1   
North Toronto West                                                     20   
Parkdale, Roncesvalles                                                 13   
Queen's Park, Ontario Provincial Government                            38   
Regent Park, Harbourfront                                              48   
Richmond, Adelaide, King                                               94   
Rosedale                                                                4   
Roselawn                                                                1   
Runnymede, Swansea                                                     39   
St. James Town                                                         77   
St. James Town, Cabbagetown                                            46   
Stn A PO Boxes                                                         95   
Studio District                                                        41   
Summerhill West, Rathnelly, South Hill, Forest ...                     17   
The Annex, North Midtown, Yorkville                                    22   
The Beaches                                                             4   
The Danforth West, Riverdale                                           43   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         36   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             57   
Brockton, Parkdale Village, Exhibition Place                            23   
Business reply mail Processing CentrE                                   18   
CN Tower, King and Spadina, Railway Lands, Harb...                      17   
Central Bay Street                                                      63   
Christie                                                                17   
Church and Wellesley                                                    78   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              36   
Davisville North                    

In [63]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 231 uniques categories.


### Analyze each neighborhood

In [64]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0            0        0                   0             0               0   
1            0        0                   0             0               0   
2            0        0                   0             0               0   
3            0        0                   0             0               0   
4            0        0                   0             0               0   

   Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0                0                 0                    0             0   
1                0                 0                    0             0   
2                0                 0                    0             0   
3                0                 0                    0             0   
4                0                 0                    0             0   

   Aquarium  ...  Theme Restaurant  Toy / Game Store  Trail  Train Station  \
0         0  ...                 0                 0      0              0   
1         0  ...                 0                 0      0              0   
2         0  ...                 0                 0      0              0   
3         0  ...                 0                 0      0              0   
4         0  ...                 0                 0      0              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Wine Bar  Wine Shop  Women's Store  
0         0          0              0  
1         0          0              0  
2         0          0              0  
3         0          0              0  
4         0          0              0  

[5 rows x 231 columns]

In [65]:
toronto_onehot.shape

(1623, 231)

In [66]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio   Airport  \
0                                         Berczy Park     0.000000  0.000000   
1        Brockton, Parkdale Village, Exhibition Place     0.000000  0.000000   
2               Business reply mail Processing CentrE     0.055556  0.000000   
3   CN Tower, King and Spadina, Railway Lands, Har...     0.000000  0.058824   
4                                  Central Bay Street     0.015873  0.000000   
5                                            Christie     0.000000  0.000000   
6                                Church and Wellesley     0.025641  0.000000   
7                      Commerce Court, Victoria Hotel     0.000000  0.000000   
8                                          Davisville     0.000000  0.000000   
9                                    Davisville North     0.000000  0.000000   
10                       Dufferin, Dovercourt Village     0.000000  0.000000   
11             First Canadian Place, Underground city     0.000000  0.000000   
12                           Forest Hill North & West     0.000000  0.000000   
13                           Garden District, Ryerson     0.000000  0.000000   
14  Harbourfront East, Union Station, Toronto Islands     0.000000  0.000000   
15                      High Park, The Junction South     0.000000  0.000000   
16                     India Bazaar, The Beaches West     0.000000  0.000000   
17          Kensington Market, Chinatown, Grange Park     0.000000  0.000000   
18                                      Lawrence Park     0.000000  0.000000   
19                           Little Portugal, Trinity     0.024390  0.000000   
20                        Moore Park, Summerhill East     0.000000  0.000000   
21                                 North Toronto West     0.050000  0.000000   
22                             Parkdale, Roncesvalles     0.000000  0.000000   
23        Queen's Park, Ontario Provincial Government     0.026316  0.000000   
24                          Regent Park, Harbourfront     0.020833  0.000000   
25                           Richmond, Adelaide, King     0.000000  0.000000   
26                                           Rosedale     0.000000  0.000000   
27                                           Roselawn     0.000000  0.000000   
28                                 Runnymede, Swansea     0.025641  0.000000   
29                                     St. James Town     0.000000  0.000000   
30                        St. James Town, Cabbagetown     0.000000  0.000000   
31                                     Stn A PO Boxes     0.010526  0.000000   
32                                    Studio District     0.024390  0.000000   
33  Summerhill West, Rathnelly, South Hill, Forest...     0.000000  0.000000   
34                The Annex, North Midtown, Yorkville     0.000000  0.000000   
35                                        The Beaches     0.000000  0.000000   
36                       The Danforth West, Riverdale     0.023256  0.000000   
37           Toronto Dominion Centre, Design Exchange     0.000000  0.000000   
38                     University of Toronto, Harbord     0.027778  0.000000   

    Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0             0.000000      0.000000        0.000000         0.000000   
1             0.000000      0.000000        0.000000         0.000000   
2             0.000000      0.000000        0.000000         0.000000   
3             0.058824      0.058824        0.117647         0.176471   
4             0.000000      0.000000        0.000000         0.000000   
5             0.000000      0.000000        0.000000         0.000000   
6             0.000000      0.000000        0.000000         0.000000   
7             0.000000      0.000000        0.000000         0.000000   
8             0.000000      0.000000        0.000000         0.000000   
9             0.000000      0.000000        0.000000         0.000000   
10            0.000000      0.000000 

Let's confirm the new size

In [67]:
toronto_grouped.shape

(39, 231)

### Let's print each neighborhood along with top 5 most common venues

In [68]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                venue  freq
0         Coffee Shop  0.07
1        Cocktail Bar  0.05
2                Café  0.04
3          Restaurant  0.04
4  Seafood Restaurant  0.04


----Brockton, Parkdale Village, Exhibition Place----
            venue  freq
0            Café  0.13
1  Breakfast Spot  0.09
2       Nightclub  0.09
3     Coffee Shop  0.09
4   Grocery Store  0.04


----Business reply mail Processing CentrE----
              venue  freq
0       Yoga Studio  0.06
1     Auto Workshop  0.06
2       Pizza Place  0.06
3        Comic Shop  0.06
4  Recording Studio  0.06


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
              venue  freq
0   Airport Service  0.18
1    Airport Lounge  0.12
2  Airport Terminal  0.12
3          Boutique  0.06
4  Sculpture Garden  0.06


----Central Bay Street----
                venue  freq
0         Coffee Shop  0.17
1  Italian Restaurant  0.08
2                Café  

#### Let's put that into a pandas dataframe

In [69]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [70]:
import numpy as np

num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2              Business reply mail Processing CentrE           Yoga Studio   
3  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar              Beer Bar           Cheese Shop   
1           Coffee Shop             Nightclub        Breakfast Spot   
2         Auto Workshop                  Park            Comic Shop   
3        Airport Lounge      Airport Terminal      Sculpture Garden   
4                  Café    Italian Restaurant        Sandwich Place   

  5th Most Common Venue      6th Most Common Venue 7th Most Common Venue  \
0            Restaurant                       Café                Bakery   
1             Pet Store                    Stadium          Intersection   
2           Pizza Place           Recording Studio         Burrito Place   
3              Boutique        Rental Car Location                 Plane   
4        Ice Cream Shop  Middle Eastern Restaurant                   Bar   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0    Seafood Restaurant                Bistro              Jazz Club  
1                Bakery    Italian Restaurant             Restaurant  
2            Restaurant               Brewery     Light Rail Station  
3         Boat or Ferry       Harbor / Marina                    Bar  
4       Thai Restaurant          Burger Joint    Fried Chicken Joint

### Cluster Neighborhoods

In [71]:
from sklearn.cluster import KMeans

# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1])

In [72]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Postal code           Borough                                 Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.654260 -79.360636               1           Coffee Shop   
1  43.662301 -79.389494               1           Coffee Shop   
2  43.657162 -79.378937               1        Clothing Store   
3  43.651494 -79.375418               1                  Café   
4  43.676357 -79.293031               0                 Trail   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Park                Bakery                   Pub   
1      Sushi Restaurant                 Diner           Yoga Studio   
2           Coffee Shop                  Café        Cosmetics Shop   
3           Coffee Shop          Cocktail Bar   American Restaurant   
4     Health Food Store                   Pub         Women's Store   

  5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
0        Breakfast Spot                  Café                      Theater   
1         Burrito Place              Beer Bar           Italian Restaurant   
2            Restaurant   Japanese Restaurant           Italian Restaurant   
3             Gastropub                 Hotel                   Restaurant   
4          Dance Studio     Electronics Store  Eastern European Restaurant   

  8th Most Common Venue      9th Most Common Venue 10th Most Common Venue  
0    Mexican Restaurant                 Restaurant         Chocolate Shop  
1             Juice Bar             Sandwich Place           Burger Joint  
2       Bubble Tea Shop  Middle Eastern Restaurant       Ramen Restaurant  
3                   Gym         Italian Restaurant       Department Store  
4            Donut Shop           Doner Restaurant                Dog Run

In [73]:
import matplotlib.cm as cm
import matplotlib.colors as colors

# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Examine Clusters

#### Cluster 1:

In [74]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
4  East Toronto               0                 Trail     Health Food Store   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
4                   Pub         Women's Store          Dance Studio   

  6th Most Common Venue        7th Most Common Venue 8th Most Common Venue  \
4     Electronics Store  Eastern European Restaurant            Donut Shop   

  9th Most Common Venue 10th Most Common Venue  
4      Doner Restaurant                Dog Run

#### Cluster 2:

In [78]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0   Downtown Toronto               1           Coffee Shop   
1   Downtown Toronto               1           Coffee Shop   
2   Downtown Toronto               1        Clothing Store   
3   Downtown Toronto               1                  Café   
5   Downtown Toronto               1           Coffee Shop   
6   Downtown Toronto               1           Coffee Shop   
7   Downtown Toronto               1         Grocery Store   
8   Downtown Toronto               1           Coffee Shop   
9       West Toronto               1                Bakery   
10  Downtown Toronto               1           Coffee Shop   
11      West Toronto               1                   Bar   
12      East Toronto               1      Greek Restaurant   
13  Downtown Toronto               1           Coffee Shop   
14      West Toronto               1                  Café   
15      East Toronto               1        Sandwich Place   
16  Downtown Toronto               1           Coffee Shop   
17      East Toronto               1                  Café   
20   Central Toronto               1                  Park   
22      West Toronto               1    Mexican Restaurant   
23   Central Toronto               1           Coffee Shop   
24   Central Toronto               1                  Café   
25      West Toronto               1             Gift Shop   
26   Central Toronto               1          Dessert Shop   
27  Downtown Toronto               1                  Café   
28      West Toronto               1           Pizza Place   
30  Downtown Toronto               1                  Café   
31   Central Toronto               1                   Pub   
32  Downtown Toronto               1       Airport Service   
34  Downtown Toronto               1           Coffee Shop   
35  Downtown Toronto               1           Coffee Shop   
36  Downtown Toronto               1           Coffee Shop   
37  Downtown Toronto               1           Coffee Shop   
38      East Toronto               1           Yoga Studio   

   2nd Most Common Venue 3rd Most Common Venue        4th Most Common Venue  \
0                   Park                Bakery                          Pub   
1       Sushi Restaurant                 Diner                  Yoga Studio   
2            Coffee Shop                  Café               Cosmetics Shop   
3            Coffee Shop          Cocktail Bar          American Restaurant   
5           Cocktail Bar              Beer Bar                  Cheese Shop   
6                   Café    Italian Restaurant               Sandwich Place   
7                   Café                  Park                  Candy Store   
8                   Café            Restaurant              Thai Restaurant   
9               Pharmacy           Supermarket                  Music Venue   
10              Aquarium                 Hotel                         Café   
11            Restaurant           Men's Store        Vietnamese Restaurant   
12           Coffee Shop    Italian Restaurant               Ice Cream Shop   
13                 Hotel                  Café                   Restaurant   
14           Coffee Shop             Nightclub               Breakfast Spot   
15  Fast Food Restaurant         Movie Theater                          Pub   
16                  Café            Restaurant                        Hotel   
17           Coffee Shop             Gastropub                       Bakery   
20      Department Store                   Gym               Breakfast Spot   
22                  Café       Thai Restaurant         Fast Food Restaurant   
23        Clothing Store                  Park           Salon / Barbershop   
24        Sandwich Place           Coffee Shop                  Pizza Place   
25      Cuban Restaurant        Breakfast Spot  Eastern European Restaurant   
26        Sandwich Place           Pizza Place                  Coffee Shop   
27      

#### Cluster 3:

In [79]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
18   Central Toronto               2           Swim School   
21   Central Toronto               2                  Park   
33  Downtown Toronto               2                  Park   

   2nd Most Common Venue       3rd Most Common Venue 4th Most Common Venue  \
18                  Park  Construction & Landscaping              Bus Line   
21         Jewelry Store                       Trail              Bus Line   
33            Playground                       Trail         Women's Store   

   5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
18         Women's Store  Ethiopian Restaurant            Electronics Store   
21      Sushi Restaurant         Deli / Bodega            Electronics Store   
33          Dance Studio     Electronics Store  Eastern European Restaurant   

          8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
18  Eastern European Restaurant            Donut Shop       Doner Restaurant  
21  Eastern European Restaurant            Donut Shop       Doner Restaurant  
33                   Donut Shop      Doner Restaurant                Dog Run

#### Cluster 4:

In [80]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
19  Central Toronto               3                Garden   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
19         Women's Store         Deli / Bodega           Event Space   

   5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
19  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
19            Donut Shop      Doner Restaurant                Dog Run

#### Cluster 5:

In [81]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
29  Central Toronto               4                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
29         Women's Store         Deli / Bodega  Ethiopian Restaurant   

   5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
29     Electronics Store  Eastern European Restaurant            Donut Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
29      Doner Restaurant               Dog Run    Distribution Center